### PR TITLE
Fix Controller:MechanicalVentilation ProportionalControlBasedOnDesignOccupancy 

### LINKED ISSUE #9969: Controller:MechanicalVentilation ProportionalControlBasedOnDesignOccupancy possible error for more than one zone
Issue overview
--------------
1. While working on #9947, noticed that `CO2PeopleGeneration` is initialized to zero at the top of `MixedAir::CalcMechVentController` is accumulated across all People objects for a given zone, but never reset to zero when looping on to the next zone. This applies only to `SystemOAMethod == DataSizing::SysOAMethod::ProportionalControlDesOcc`.
2. The [accumulating loop](https://github.com/NREL/EnergyPlus/blob/a116efebd3867cf0eaabfa8e60c6210e7b6992fd/src/EnergyPlus/MixedAir.cc#L3801-L3806) is inefficient, because it loops over *all* people statements for every zone on the controller looking for a matching zone number.
3. This loop is not covered by integration tests.
4. This is kind of covered by unit test `CO2ControlDesignOccupancyTest` except the CO2 values are zero, because the test setup is incomplete (People activity schedule and CO2 generation rate are both zero).
5. The documentation for Controller:MechanicalVentilation System Outdoor Air Method is not completely clear, implies this option is only for single-zone system, and  this could use some reformatting to make it more readable.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 23.1
 - Related to #9947

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #9969
 - Streamlines calculation of current CO2 by storing list of applicable people statements instead of looping through all people statements for every vent mech zone every timestep looking for matching zone numbers.
 - Revised single-zone unit test adds several cases for current zone CO2 above, below, and in between the zone max/min CO2 values. The revised test passes both here and in develop.
 - Second unit test has three identical zones, so 3x OA flow rate is expected. This test passes here, but fails in develop. See https://github.com/NREL/EnergyPlus/pull/10268#issuecomment-1771539050

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mjwitte** [comment]
Single-zone unit test passes in develop, but 3-zone unit test fails in develop.
```
[----------] 2 tests from EnergyPlusFixture
[ RUN      ] EnergyPlusFixture.CO2ControlDesignOccupancyTest
[       OK ] EnergyPlusFixture.CO2ControlDesignOccupancyTest (14693 ms)
[ RUN      ] EnergyPlusFixture.CO2ControlDesignOccupancyTest3Zone
C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\MixedAir.unit.cc(1263): error: The difference between 3 * expectedOAMassFlow and oaController.OAMassFlow is 0.0039677399219509683, which exceeds 0.00001, where
3 * expectedOAMassFlow evaluates to 0.024854760000000004,
oaController.OAMassFlow evaluates to 0.020887020078049035, and
0.00001 evaluates to 1.0000000000000001e-05.
C:\MJWGit\EnergyPlus\tst\EnergyPlus\unit\MixedAir.unit.cc(1264): error: The difference between 3 * expectedOAMassFlow / oaController.MixMassFlow and oaController.MinOAFracLimit is 0.0019449705499759648, which exceeds 0.00001, where
3 * expectedOAMassFlow / oaController.MixMassFlow evaluates to 0.012183705882352942,
oaController.MinOAFracLimit evaluates to 0.010238735332376977, and
0.00001 evaluates to 1.0000000000000001e-05.
[  FAILED  ] EnergyPlusFixture.CO2ControlDesignOccupancyTest3Zone (37483 ms)
[----------] 2 tests from EnergyPlusFixture (52178 ms total)
```

File: src/EnergyPlus/MixedAir.cc [hunk 1]
```diff
@@ -2824,6 +2824,18 @@ void InitOAController(EnergyPlusData &state, int const OAControllerNum, bool con
                 }
             }
 
+            // Fill People index lists if needed
+            if (vent_mech.SystemOAMethod == DataSizing::SysOAMethod::ProportionalControlDesOcc) {
+                for (int peopleNum = 1; peopleNum <= state.dataHeatBal->TotPeople; ++peopleNum) {
+                    for (auto &thisMechVentZone : vent_mech.VentMechZone) {
+                        if (state.dataHeatBal->People(peopleNum).ZonePtr == thisMechVentZone.zoneNum) {
+                            thisMechVentZone.peopleIndexes.push_back(peopleNum);
```
Comment (mjwitte) [inline src/EnergyPlus/MixedAir.cc:2824]: Save list of applicable people instances.

File: src/EnergyPlus/MixedAir.cc [hunk 2]
```diff
@@ -3672,19 +3684,18 @@ Real64 VentilationMechanicalProps::CalcMechVentController(EnergyPlusData &state,
     // new local variables for DCV
     // Zone OA flow rate based on each calculation method [m3/s]
     std::array<Real64, static_cast<int>(DataSizing::OAFlowCalcMethod::Num)> ZoneOACalc{0.0};
-    Real64 CO2PeopleGeneration = 0; // CO2 generation from people at design level
```
Comment (mjwitte) [inline src/EnergyPlus/MixedAir.cc:3664]: Narrow scope for this.

File: src/EnergyPlus/MixedAir.cc [hunk 3]
```diff
@@ -3946,8 +3957,8 @@ Real64 VentilationMechanicalProps::CalcMechVentController(EnergyPlusData &state,
                                         // Calculate zone maximum target CO2 concentration in PPM
                                         if (this->SystemOAMethod == DataSizing::SysOAMethod::ProportionalControlDesOcc) {
                                             // Accumulate CO2 generation from people at design occupancy and current activity level
-                                            for (int PeopleNum = 1; PeopleNum <= state.dataHeatBal->TotPeople; ++PeopleNum) {
-                                                if (state.dataHeatBal->People(PeopleNum).ZonePtr != ZoneNum) continue;
+                                            Real64 CO2PeopleGeneration = 0.0;
+                                            for (int const PeopleNum : thisMechVentZone.peopleIndexes) {
```
Comment (mjwitte) [inline src/EnergyPlus/MixedAir.cc:3950]: Loop through only the applicable people instances, not all of them. Also reset sum to zero for each zone (previously this kept accumulating across zones).

File: testfiles/CMakeLists.txt [hunk 4]
```diff
@@ -355,6 +355,7 @@ add_simulation_test(IDF_FILE HeatPumpIAQP_DCV.idf EPW_FILE USA_IL_Chicago-OHare.
 add_simulation_test(IDF_FILE HeatPumpIAQP_GenericContamControl.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE HeatPumpProportionalControl_DCV.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE HeatPumpProportionalControl_DCVDesignRate.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+add_simulation_test(IDF_FILE HeatPumpProportionalControl_DCVDesignOccAllZones.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
```
Comment (mjwitte) [inline testfiles/CMakeLists.txt:358]: Add a test file that uses this control type.

**Review by mjwitte — COMMENTED** [review]
Code walkthru.

File: tst/EnergyPlus/unit/MixedAir.unit.cc [hunk 5]
```diff
@@ -874,14 +912,359 @@ TEST_F(EnergyPlusFixture, CO2ControlDesignOccupancyTest)
...
+    state->dataEnvrn->StdRhoAir = 1.2;
+    oaController.MixMassFlow = 1.7 * state->dataEnvrn->StdRhoAir;
+    oaController.MaxOAMassFlowRate = 1.7 * state->dataEnvrn->StdRhoAir;
+    state->dataAirLoop->AirLoopFlow(1).DesSupply = 1.7 * state->dataEnvrn->StdRhoAir;
+
+    // From peeking inside the CO2 limit calcs
+    Real64 zoneCO2Max = 431.08678;
+    Real64 zoneCO2Min = 400.0;
+
+    // Case 1 - Zone CO2 greater than CO2 Max, so OA flow is flow/area+flow/person
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/MixedAir.unit.cc:926]: Single-zone test, three cases.

File: tst/EnergyPlus/unit/MixedAir.unit.cc [hunk 6]
```diff
@@ -874,14 +912,359 @@ TEST_F(EnergyPlusFixture, CO2ControlDesignOccupancyTest)
...
+    state->dataContaminantBalance->ZoneAirCO2(1) = zoneCO2Min + 0.3 * (zoneCO2Max - zoneCO2Min);
+    expectedOAMassFlow = (oaRequirements.OAFlowPerArea * state->dataHeatBal->Zone(1).FloorArea +
+                          0.3 * oaRequirements.OAFlowPerPerson * state->dataHeatBal->Zone(1).TotOccupants) *
+                         state->dataEnvrn->StdRhoAir;
+    oaController.CalcOAController(*state, 1, true);
+    EXPECT_NEAR(expectedOAMassFlow, oaController.OAMassFlow, 0.00001);
+    EXPECT_NEAR(expectedOAMassFlow / oaController.MixMassFlow, oaController.MinOAFracLimit, 0.00001);
+}
+
+TEST_F(EnergyPlusFixture, CO2ControlDesignOccupancyTest3Zone)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/MixedAir.unit.cc:952]: 3-zone test, identical to single-zone test but with 3 zones, expect 3x OA flow rates.

**Comment by Myoldmopar** [comment]
The Windows build failure is unrelated, as is the one small diff that pops up occasionally.  I'll pull develop in and build while looking this over.

File: src/EnergyPlus/MixedAir.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/MixedAir.cc:3664]: That's a great thing.

File: src/EnergyPlus/MixedAir.hh [hunk 7]
```diff
@@ -286,6 +286,7 @@ namespace MixedAir {
         DataSizing::OAFlowCalcMethod ZoneOAFlowMethod = DataSizing::OAFlowCalcMethod::PerPerson; // OA flow method for each zone
         int ZoneOASchPtr = 0;              // Index to the outdoor air schedule for each zone (from DesignSpecification:OutdoorAir or default)
         Real64 OAPropCtlMinRateSchPtr = 0; // Outdoor design OA flow rate schedule from DesignSpecification:OutdoorAir
+        EPVector<int> peopleIndexes; // List of People objects in this zone (for SystemOAMethod == DataSizing::SysOAMethod::ProportionalControlDesOcc)
```
Comment (Myoldmopar) [inline src/EnergyPlus/MixedAir.hh:289]: Yay, a nice little EPVector!

File: testfiles/CMakeLists.txt (hunk 4 above)
Comment (Myoldmopar) [reply testfiles/CMakeLists.txt:358]: Sounds good to me.

**Review by Myoldmopar — APPROVED** [review]
Looks like fine little fix and cleanup with sufficient tests to me.

**Comment by Myoldmopar** [comment]
Everything tests out perfectly fine, and the changes look good.  Merging this.  Thanks @mjwitte

**Pull request merged**